Trib 73: checks if phrase has been reviewed, and if it has, applies to user

mrsbluerose opened the pull request:
**4/17/23: Pull request to check approach -- not complete** 
Tested successful in Postman - checks if the phrase exists and applies to user (I used an existing phrase for user 1 and applied to user 2)

I took some liberties with the database, and I'm not sure this is the approach you wanted. Here are all of my notes. Please let me know if you want changes, a different approach or if I should continue and button this up:

- Left some test print statements I'm using while working through this and testing
- Updated word methods and HasBeenReviewed method in PhraseService to check for words and return ids (instead of boolean)
- Updated AttributesAPIController call and Phrase service methods to include a user Id. In Postman, I included "id" in the body JSON.
- Added migration script to make userId and phraseId a composite primary key for the user_phrase table
- Updated UserPhrase entity to include the composite key and entity notation. This was the result of troubleshooting attempts to save the user and phrase ids in the UserPhraseRepo. Used reference: [https://www.codejava.net/frameworks/spring/spring-data-jpa-composite-primary-key-examples]. This included:
1. Adding a UserPhraseId class. I was not sure where to put this. It's currently in the entities directory.
2. Adding the @Entity, @ID and @IdClass notations to UserPhrase entity (also added getters and setters for when it's used to save to repo)
3. Updating UserPhraseRepository to use parameters <UserPhrase, UserPhraseId>

**Remaining things to address:**
- I was planning to add logic to check that the phrase isn't already mapped to the user before adding to the user_phrase table.
- Question: The applyPhraseToUser methods are currently void type. Do we want to return any kind of success message or indication of what happened (was applied, was not applied, already applied)?

mrsbluerose reviewed (commented):
@haxwell when I updated the req.id in this call to req.userId, it did change the AttributeRequestDTO. I noticed that the AttributeRequestDTO is actually specified in the endpoint signature:
`@RequestMapping(value = { "/api/attributes" }, method=RequestMethod.POST)
    public ResponseEntity<Boolean> applyPhraseToUser(@RequestBody @Valid AttributesRequest req) {`
This is contrary to what we discussed in the last meeting, where you said that AttributeRequstDTO should be for responses and the id field is for the attribute id. I think this may have been overlooked when I worked on separating "attributes" and "Phrase" functionality in TRIB-33. How do you want this to work? Any changes, or should I leave it as is?

mrsbluerose commented on src/main/java/com/savvato/tribeapp/controllers/AttributesAPIController.java:
```diff
@@ -45,7 +45,7 @@ public ResponseEntity<Boolean> applyPhraseToUser(@RequestBody @Valid AttributesR
 
         if (phraseService.isPhraseValid(req.adverb, req.verb, req.preposition, req.noun)) {
-            phraseService.applyPhraseToUser(req.id, req.adverb, req.verb, req.preposition, req.noun);
+            phraseService.applyPhraseToUser(req.userId, req.adverb, req.verb, req.preposition, req.noun);
```
@haxwell when I updated the req.id in this call to req.userId, it did change the AttributeRequestDTO. I noticed that the AttributeRequestDTO is actually specified in the endpoint signature:
`@RequestMapping(value = { "/api/attributes" }, method=RequestMethod.POST) public ResponseEntity<Boolean> applyPhraseToUser(@RequestBody @Valid AttributesRequest req) {`
This is contrary to what we discussed in the last meeting, where you said that AttributeRequstDTO should be for responses and the id field is for the attribute id. I think this may have been overlooked when I worked on separating "attributes" and "phrase" functionality in TRIB-33. How do you want this to work? Any changes, or should I leave it as is?

haxwell replied:
@mrsbluerose The *Request objects describe the possible fields that the JSON body of a request to an associated endpoint might contain.

So AttributesRequest's id field would be the ID of the attribute the client was concerned with.

The response objects end in DTO. They describe the data returned in response to an endpoint call.

So, I think the way to resolve this is to add a field to the AttributesRequest for 'userId'. Keep the 'id' field, and add a 'userId' field.

haxwell commented on src/main/java/com/savvato/tribeapp/controllers/AttributesAPIController.java:
```diff
@@ -41,11 +41,10 @@ public ResponseEntity<List<AttributeDTO>> getAttributesForUser(@PathVariable Lon
-
         if (phraseService.isPhraseValid(req.adverb, req.verb, req.preposition, req.noun)) {
-            phraseService.applyPhraseToUser(req.adverb, req.verb, req.preposition, req.noun);
+            System.out.println("Controller: phrase is valid"); //////
```
Another thought: most of the `log` messages defined in this PR will be `DEBUG`. Do a bit of googling, and be sure you have an idea when to use `INFO`, `WARNING` and `ERROR`, the other log statuses.

mrsbluerose replied:
Replied below with an update of all my changes. Added Loggers (warning and info). I removed everything else.  After looking through them, I felt like the remaining print statements were really just for my info as I was working through things and not necessary otherwise. I also threw an illegal exception if the verb or noun were blank (let me know if you want that to be a warning instead).

mrsbluerose commented:
Updates:

- refactored isPhraseValid to check for empty/null verb/noun, rejected words and rejected phrase
- added logic to convert words to lower case to isPhraseValid and applyPhraseToUser
- refactored findPreviouslyApprovedPhraseId to return result sooner if any single word is not in our word databases 
- refactored isAnyWordRejected to return result sooner if any word is found (also refactored name and result to better match what we want to find - isAnyWordRejected (true if they are and false if they are not)
- removed unnecessary comments and lines
- added Logger messages
- added unit tests for isPhraseValid

The ticket conditions are complete (minus any changes you want)

I would like to eventually add a unit test for applyPhraseToUser. I researched capturing log messages and could not get a solution to work.

mrsbluerose commented on src/main/java/com/savvato/tribeapp/services/PhraseServiceImpl.java:
```diff
@@ -31,60 +29,170 @@ public class PhraseServiceImpl implements PhraseService {
+    UserPhraseRepository userPhraseRepository;
+
+
+    //isPhraseValid (call following methods)
```
cleaned it all up

haxwell reviewed (commented):
well done!
